common/phal_F4_F7/usart/usart.h

```diff
@@ -0,0 +1,265 @@
+    uint8_t _tx_busy;              //!< Waiting on a transmission to finish
+    volatile uint8_t _rx_busy;     //!< Waiting on a reception to finish
+    volatile uint32_t rxfer_size;  //!< Size of data to receive over DMA
+} usart_active_transfer_t;
+
+// Function Prototypes
+
+/**
+ * @brief Initialize a USART Peripheral with desired settings
+ *
+ * @param handle Handle containing settings for USART peripheral
+ * @param fck Clock rate going to USART peripheral (APBx bus)
+ * @return true Successfully initialized USART peripheral
+ * @return false Failed to initialize USART peripheral
+ */
+bool PHAL_initUSART(usart_init_t* handle, const uint32_t fck);
+
+
+/**
+ * @brief           Starts a tx using dma, use PHAL_usartTxDmaComplete
+ *                  to ensure the previous transmission is complete
+ * @param handle    The handle for the usart configuration
+ * @param data      The address of the data to send, ensure a cast to (uint16_t *), even if 8 bits
+ * @param len       Number of units of data, depending on the configured word length
+ */
+bool PHAL_usartTxDma(usart_init_t* handle, uint16_t* data, uint32_t len);
+
+/**
+ * @brief           Starts an rx using dma of a specific length
+ *
+ * @param handle    The handle for the usart configuration
+ * @param data      The address to put the received data, ensure a cast to (uint16_t *), even if 8 bits
+ * @param len       Number of units of data, depending on the configured word length
+ * @param cont      Enable Continous RX using the idle line interrupt (only need to call this function once, and HAL will keep recieving messages of the same length)
+ */
+bool PHAL_usartRxDma(usart_init_t* handle, uint16_t* data, uint32_t len, bool cont);
+
+/**
+ * @brief Disables the Continous RX that was previously used
+ *
+ * @param handle The handle for the usart configuration
+ */
+bool PHAL_disableContinousRxDMA(usart_init_t* handle);
+
+/**
+ * @brief Returns whether USART peripheral is currently transmitting data
+ *
+ * @param handle Handle of USART peripheral to check
+ * @return true USART peripheral is currently sending a message
+ * @return false USART peripheal is not currently sending a message
+ */
+bool PHAL_usartTxBusy(usart_init_t* handle);
+
+/**
+ * @brief Returns whether USART peripheral is currently receiving data
+ *
+ * @param handle Handle of USART peripheral check
+ * @return true USART peripheral is currently receiving a message
+ * @return false USART peripheal is not currently receiving a message
+ */
+bool PHAL_usartRxBusy(usart_init_t *handle);
+
+/**
+ * @brief Callback function called immediately after reception of a USART RX message
+ * Uses USART IDLE line interrupt
+ *
+ * NOTE: this is executed during an interrupt handler call, so keep code in this function light
+ *
+ * @param handle Handle of USART peripheral that just recieved a message
+ */
+extern void usart_recieve_complete_callback(usart_init_t *handle);
```

Comment:
If this works it will greatly simplify parsing

Reply:
It does!